### PR TITLE
BLD: Don't assume -pthread is already set on compiler

### LINKED ISSUE #11323: Compilation failure due to missing symbol pthread_atfork
#10614 introduced usage of `pthread_atfork` if the pthread header was found. However this requires adding `-lpthread` or `-pthread` to the compiler flags so the symbols can be found.

This kinda works on x86 platforms because Python is usually build with pthreads already and the dynamic library contains all the symbols. However on e.g. Power9 `libpthread` is a linker script contained a static part which contains e.g. `pthread_atfork` which does now not get linked and hence usage fails with 

  ```
File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/__init__.py", line 156, in <module>
    from . import fft
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/__init__.py", line 81, in <module>
    from ._helper import next_fast_len
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_helper.py", line 4, in <module>
    from . import _pocketfft
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/__init__.py", line 3, in <module>
    from .basic import *
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/basic.py", line 8, in <module>
    from . import pypocketfft as pfft
ImportError: /tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-37m-powerpc64le-linux-gnu.so: undefined symbol: pthread_atfork
 (at easybuild/tools/run.py:529 in parse_cmd_output)
```

Proposed solution: Add `-pthread` or `-lpthread` to the link flags.

### PR DESCRIPTION
**Pull request by peterbell10**
#### Reference issue
Fixes gh-11323

#### What does this implement/fix?
This explicitly adds the `-pthread` argument to `pypocketfft` builds. This isn't usually required since `distutils` usually adds this argument already.

### CODE REVIEW

File: scipy/fft/_pocketfft/setup.py [hunk 1]
```diff
@@ -5,20 +5,28 @@ def pre_build_hook(build_ext, ext):
...
                                    'int main(int argc, char **argv) {}')
         if has_pthreads:
             ext.define_macros.append(('POCKETFFT_PTHREADS', None))
-
+            if has_flag(cc, '-pthread'):
+                args.append('-pthread')
+                ext.extra_link_args.append('-pthread')
+            else:
+                from numpy.distutils import log
+                log.warn('Unknown compiler pthread flag')
```
Comment (Flamefire) [inline scipy/fft/_pocketfft/setup.py]: I'd make this a hard error instead. It is really cumbersome to build this and later find that it doesn't work. My current approach I'm testing is `ext.libraries.append('pthread')` (unconditionally)

The other is to try_link

            pthread_code = """#include <pthread.h>
                        void* thread_start(void* data){ return data; }
                        void dummy(){}
                        int main() {
                            pthread_t thread;
                            pthread_create(&thread, 0, thread_start, 0);
                            pthread_atfork(dummy, dummy, dummy);
                        }"""

and try `('', '-pthread', '-lpthread', '-lpthreads', '-lposix', '-lthread')` till one works

**Comment by Flamefire** [comment]
I successfully tested this on the system it failed previously. Thanks!

**Review by Flamefire — APPROVED** [review]
Approved as I successfully installed SciPy with this patch on Power9 where it failed before

**Review by larsoner — APPROVED** [review]

**Comment by tylerjereddy** [comment]
Alright, original issue reporter confirmed fix & there's a core dev approval. FWIW, I don't observe any build issues like this on POWER8 or POWER9 gcc compile farm machines using latest master, so it seems to apply to that very specific build scenario with certain flags set, etc.

**Comment by tylerjereddy** [comment]
Thanks @peterbell10

**Pull request merged**